File: packages/auth/src/messages.ts
import type { NativeAuthorityUnavailableReason } from "@recurs/contracts";

import {
  NativeMessageType,
  allocateNativeBytes,
  copyNativeBytes,
  encodeNativeFrame,
  failNativeCodec,
} from "./frame.js";
import type { NativeFrame } from "./frame.js";
import {
  decodeBoolean,
  decodeFieldTable,
  decodeNonce,
  decodeU16,
  decodeVersionText,
  encodeFieldTable,
  encodeNonce,
  encodeU32,
  encodeVersionText,
} from "./fields.js";
import type { NativeField } from "./fields.js";

export interface NativeHello {
  readonly engineVersion: string;
  readonly nonce: Uint8Array;
}

export interface NativeHelloResult {
  readonly launcherVersion: string;
  readonly brokerVersion: string;
  readonly echoedNonce: Uint8Array;
  readonly productionSigned: boolean;
  readonly persistentCredentials: boolean;
  readonly minimumMacosVersion: "14.4";
}

export interface NativeHealthResult {
  readonly keychain: "available" | "locked" | "unavailable";
  readonly peerVerified: boolean;
}

export enum NativeKeychainStatusCode {
  available = 1,
  locked = 2,
  unavailable = 3,
}
Object.freeze(NativeKeychainStatusCode);

export enum NativeSafeFailureCode {
  unsupportedPlatform = 1,
  unsupportedOsVersion = 2,
  launcherUnavailable = 3,
  brokerUnavailable = 4,
  protocolMismatch = 5,
  peerIdentityUnverified = 6,
  productionSigningRequired = 7,
  keychainUnavailable = 8,
  unsupportedOperation = 9,
}
Object.freeze(NativeSafeFailureCode);

function requireExactFields(
  frame: NativeFrame,
  expectedType: NativeMessageType,
  expectedTags: readonly number[],
): readonly NativeField[] {
  if (frame.type !== expectedType) {
    failNativeCodec("invalid_message");
  }
  const fields = decodeFieldTable(frame.payload);
  if (
    fields.length !== expectedTags.length ||
    fields.some((field, index) => field.tag !== expectedTags[index])
  ) {
    failNativeCodec("invalid_message");
  }
  return fields;
}

function decodeSemanticMessage<T>(decode: () => T): T {
  try {
    return decode();
  } catch {
    failNativeCodec("invalid_message");
  }
}

export function encodeHello(requestId: number, hello: NativeHello): Uint8Array {
  let engineVersion: string;
  let nonce: Uint8Array;
  try {
    engineVersion = hello.engineVersion;
    nonce = hello.nonce;
  } catch {
    failNativeCodec("invalid_message");
  }
  const payload = encodeFieldTable([
    { tag: 1, value: encodeVersionText(engineVersion) },
    { tag: 2, value: encodeNonce(nonce) },
  ]);
  return encodeNativeFrame({
    type: NativeMessageType.hello,
    requestId,
    payload,
  });
}

export function decodeHelloResult(frame: NativeFrame): NativeHelloResult {
  return decodeSemanticMessage(() => {
    const fields = requireExactFields(
      frame,
      NativeMessageType.helloResult,
      [1, 2, 3, 4, 5, 6],
    );
    const launcherVersion = decodeVersionText(
      fields[0]?.value ?? allocateNativeBytes(0),
    );
    const brokerVersion = decodeVersionText(
      fields[1]?.value ?? allocateNativeBytes(0),
    );
    const echoedNonce = decodeNonce(
      fields[2]?.value ?? allocateNativeBytes(0),
    );
    const productionSigned = decodeBoolean(
      fields[3]?.value ?? allocateNativeBytes(0),
    );
    const persistentCredentials = decodeBoolean(
      fields[4]?.value ?? allocateNativeBytes(0),
    );
    const minimumMacosVersion = decodeVersionText(
      fields[5]?.value ?? allocateNativeBytes(0),
    );
    if (minimumMacosVersion !== "14.4") {
      failNativeCodec("invalid_message");
    }

    const storedNonce = copyNativeBytes(echoedNonce);
    return Object.freeze({
      launcherVersion,
      brokerVersion,
      get echoedNonce(): Uint8Array {
        return copyNativeBytes(storedNonce);
      },
      productionSigned,
      persistentCredentials,
      minimumMacosVersion,
    });
  });
}

export function encodeHealth(requestId: number): Uint8Array {
  return encodeNativeFrame({
    type: NativeMessageType.health,
    requestId,
    payload: encodeFieldTable([]),
  });
}

const keychainValues = Object.freeze({
  [NativeKeychainStatusCode.available]: "available",
  [NativeKeychainStatusCode.locked]: "locked",
  [NativeKeychainStatusCode.unavailable]: "unavailable",
} as const);

export function decodeHealthResult(frame: NativeFrame): NativeHealthResult {
  return decodeSemanticMessage(() => {
    const fields = requireExactFields(
      frame,
      NativeMessageType.healthResult,
      [1, 2],
    );
    const keychainCode = decodeU16(
      fields[0]?.value ?? allocateNativeBytes(0),
    );
    const keychain = keychainValues[keychainCode as keyof typeof keychainValues];
    if (keychain === undefined) {
      failNativeCodec("invalid_message");
    }
    const peerVerified = decodeBoolean(
      fields[1]?.value ?? allocateNativeBytes(0),
    );
    return Object.freeze({ keychain, peerVerified });
  });
}

export function encodeCancel(
  requestId: number,
  targetRequestId: number,
): Uint8Array {
  if (
    !Number.isInteger(targetRequestId) ||
    targetRequestId <= 0 ||
    targetRequestId > 0xffff_ffff
  ) {
    failNativeCodec("invalid_message");
  }
  return encodeNativeFrame({
    type: NativeMessageType.cancel,
    requestId,
    payload: encodeFieldTable([{ tag: 1, value: encodeU32(targetRequestId) }]),
  });
}

const safeFailureValues: Readonly<Record<number, NativeAuthorityUnavailableReason>> =
  Object.freeze({
    [NativeSafeFailureCode.unsupportedPlatform]: "unsupported_platform",
    [NativeSafeFailureCode.unsupportedOsVersion]: "unsupported_os_version",
    [NativeSafeFailureCode.launcherUnavailable]: "launcher_unavailable",
    [NativeSafeFailureCode.brokerUnavailable]: "broker_unavailable",
    [NativeSafeFailureCode.protocolMismatch]: "protocol_mismatch",
    [NativeSafeFailureCode.peerIdentityUnverified]: "peer_identity_unverified",
    [NativeSafeFailureCode.productionSigningRequired]:
      "production_signing_required",
    [NativeSafeFailureCode.keychainUnavailable]: "keychain_unavailable",
    [NativeSafeFailureCode.unsupportedOperation]: "unsupported_operation",
  });

export function decodeSafeFailure(
  frame: NativeFrame,
): NativeAuthorityUnavailableReason {
  return decodeSemanticMessage(() => {
    const fields = requireExactFields(
      frame,
      NativeMessageType.safeFailure,
      [1],
    );
    const code = decodeU16(fields[0]?.value ?? allocateNativeBytes(0));
    const reason = safeFailureValues[code];
    if (reason === undefined) {
      failNativeCodec("invalid_message");
    }
    return reason;
  });
}

File: native/macos/Sources/RecursNativeProtocol/Messages.swift
import Foundation

public enum KeychainStatusCode: UInt16, CaseIterable, Sendable {
  case available = 1
  case locked = 2
  case unavailable = 3
}

public enum SafeFailureCode: UInt16, CaseIterable, Sendable {
  case unsupportedPlatform = 1
  case unsupportedOSVersion = 2
  case launcherUnavailable = 3
  case brokerUnavailable = 4
  case protocolMismatch = 5
  case peerIdentityUnverified = 6
  case productionSigningRequired = 7
  case keychainUnavailable = 8
  case unsupportedOperation = 9
}

public struct HelloMessage: Equatable, Sendable {
  public let engineVersion: String

  private let nonceStorage: [UInt8]

  public var nonce: Data {
    Data(nonceStorage)
  }

  public init(engineVersion: String, nonce: Data) throws {
    _ = try FieldTable.encodeVersionText(engineVersion)
    let storedNonce = try FieldTable.encodeNonce(nonce)
    self.engineVersion = engineVersion
    self.nonceStorage = Array(storedNonce)
  }

  private init(
    validatedEngineVersion engineVersion: String,
    nonceBytes: [UInt8]
  ) {
    self.engineVersion = engineVersion
    self.nonceStorage = Array(nonceBytes)
  }

  public static func decode(_ frame: NativeFrame) throws -> HelloMessage {
    try withInvalidMessage {
      let fields = try requireFields(frame, type: .hello, tags: [1, 2])
      let engineVersion = try FieldTable.decodeVersionText(fields[0].value)
      let nonce = try FieldTable.decodeNonce(fields[1].value)
      return HelloMessage(
        validatedEngineVersion: engineVersion,
        nonceBytes: Array(nonce)
      )
    }
  }

  public func encodedFrame(requestID: UInt32) throws -> Data {
    let table = try FieldTable(fields: [
      FieldTable.Field(
        tag: 1,
        value: try FieldTable.encodeVersionText(engineVersion)
      ),
      FieldTable.Field(tag: 2, value: Data(nonceStorage)),
    ])
    return try NativeFrame(
      type: .hello,
      requestID: requestID,
      payload: table.encoded()
    ).encoded()
  }
}

public struct HelloResultMessage: Equatable, Sendable {
  public let launcherVersion: String
  public let brokerVersion: String
  public let productionSigned: Bool
  public let persistentCredentials: Bool
  public let minimumMacosVersion: String

  private let echoedNonceStorage: [UInt8]

  public var echoedNonce: Data {
    Data(echoedNonceStorage)
  }

  public init(
    launcherVersion: String,
    brokerVersion: String,
    echoedNonce: Data,
    productionSigned: Bool,
    persistentCredentials: Bool,
    minimumMacosVersion: String
  ) throws {
    guard minimumMacosVersion == "14.4" else {
      throw NativeProtocolError.invalidMessage
    }
    _ = try FieldTable.encodeVersionText(launcherVersion)
    _ = try FieldTable.encodeVersionText(brokerVersion)
    _ = try FieldTable.encodeVersionText(minimumMacosVersion)
    let storedNonce = try FieldTable.encodeNonce(echoedNonce)

    self.launcherVersion = launcherVersion
    self.brokerVersion = brokerVersion
    self.echoedNonceStorage = Array(storedNonce)
    self.productionSigned = productionSigned
    self.persistentCredentials = persistentCredentials
    self.minimumMacosVersion = minimumMacosVersion
  }

  private init(
    validatedLauncherVersion launcherVersion: String,
    brokerVersion: String,
    echoedNonceBytes: [UInt8],
    productionSigned: Bool,
    persistentCredentials: Bool
  ) {
    self.launcherVersion = launcherVersion
    self.brokerVersion = brokerVersion
    self.echoedNonceStorage = Array(echoedNonceBytes)
    self.productionSigned = productionSigned
    self.persistentCredentials = persistentCredentials
    self.minimumMacosVersion = "14.4"
  }

  public static func decode(_ frame: NativeFrame) throws -> HelloResultMessage {
    try withInvalidMessage {
      let fields = try requireFields(
        frame,
        type: .helloResult,
        tags: [1, 2, 3, 4, 5, 6]
      )
      let launcherVersion = try FieldTable.decodeVersionText(fields[0].value)
      let brokerVersion = try FieldTable.decodeVersionText(fields[1].value)
      let echoedNonce = try FieldTable.decodeNonce(fields[2].value)
      let productionSigned = try FieldTable.decodeBoolean(fields[3].value)
      let persistentCredentials = try FieldTable.decodeBoolean(fields[4].value)
      let minimumMacosVersion = try FieldTable.decodeVersionText(fields[5].value)
      guard minimumMacosVersion == "14.4" else {
        throw NativeProtocolError.invalidMessage
      }

      return HelloResultMessage(
        validatedLauncherVersion: launcherVersion,
        brokerVersion: brokerVersion,
        echoedNonceBytes: Array(echoedNonce),
        productionSigned: productionSigned,
        persistentCredentials: persistentCredentials
      )
    }
  }

  public func encodedFrame(requestID: UInt32) throws -> Data {
    let table = try FieldTable(fields: [
      FieldTable.Field(
        tag: 1,
        value: try FieldTable.encodeVersionText(launcherVersion)
      ),
      FieldTable.Field(
        tag: 2,
        value: try FieldTable.encodeVersionText(brokerVersion)
      ),
      FieldTable.Field(tag: 3, value: Data(echoedNonceStorage)),
      FieldTable.Field(tag: 4, value: FieldTable.encodeBoolean(productionSigned)),
      FieldTable.Field(tag: 5, value: FieldTable.encodeBoolean(persistentCredentials)),
      FieldTable.Field(
        tag: 6,
        value: try FieldTable.encodeVersionText(minimumMacosVersion)
      ),
    ])
    return try NativeFrame(
      type: .helloResult,
      requestID: requestID,
      payload: table.encoded()
    ).encoded()
  }
}

public struct HealthResultMessage: Equatable, Sendable {
  public let keychain: KeychainStatusCode
  public let peerVerified: Bool

  public init(keychain: KeychainStatusCode, peerVerified: Bool) {
    self.keychain = keychain
    self.peerVerified = peerVerified
  }

  public static func decode(_ frame: NativeFrame) throws -> HealthResultMessage {
    try withInvalidMessage {
      let fields = try requireFields(frame, type: .healthResult, tags: [1, 2])
      let rawKeychain = try FieldTable.decodeUInt16(fields[0].value)
      guard let keychain = KeychainStatusCode(rawValue: rawKeychain) else {
        throw NativeProtocolError.invalidMessage
      }
      let peerVerified = try FieldTable.decodeBoolean(fields[1].value)
      return HealthResultMessage(keychain: keychain, peerVerified: peerVerified)
    }
  }

  public func encodedFrame(requestID: UInt32) throws -> Data {
    let table = try FieldTable(fields: [
      FieldTable.Field(tag: 1, value: FieldTable.encodeUInt16(keychain.rawValue)),
      FieldTable.Field(tag: 2, value: FieldTable.encodeBoolean(peerVerified)),
    ])
    return try NativeFrame(
      type: .healthResult,
      requestID: requestID,
      payload: table.encoded()
    ).encoded()
  }
}

public func makeHealthFrame(requestID: UInt32) throws -> Data {
  let table = try FieldTable(fields: [])
  return try NativeFrame(
    type: .health,
    requestID: requestID,
    payload: table.encoded()
  ).encoded()
}

public func makeCancelFrame(
  requestID: UInt32,
  targetRequestID: UInt32
) throws -> Data {
  guard targetRequestID != 0 else {
    throw NativeProtocolError.invalidMessage
  }
  let table = try FieldTable(fields: [
    FieldTable.Field(tag: 1, value: FieldTable.encodeUInt32(targetRequestID))
  ])
  return try NativeFrame(
    type: .cancel,
    requestID: requestID,
    payload: table.encoded()
  ).encoded()
}

extension SafeFailureCode {
  public static func decode(_ frame: NativeFrame) throws -> SafeFailureCode {
    try withInvalidMessage {
      let fields = try requireFields(frame, type: .safeFailure, tags: [1])
      let rawCode = try FieldTable.decodeUInt16(fields[0].value)
      guard let code = SafeFailureCode(rawValue: rawCode) else {
        throw NativeProtocolError.invalidMessage
      }
      return code
    }
  }

  public func encodedFrame(requestID: UInt32) throws -> Data {
    let table = try FieldTable(fields: [
      FieldTable.Field(tag: 1, value: FieldTable.encodeUInt16(rawValue))
    ])
    return try NativeFrame(
      type: .safeFailure,
      requestID: requestID,
      payload: table.encoded()
    ).encoded()
  }
}

private func requireFields(
  _ frame: NativeFrame,
  type: NativeMessageType,
  tags: [UInt16]
) throws -> [FieldTable.Field] {
  guard frame.type == type else {
    throw NativeProtocolError.invalidMessage
  }
  let fields = try FieldTable.decode(frame.payload).fields
  guard fields.count == tags.count else {
    throw NativeProtocolError.invalidMessage
  }
  for (field, expectedTag) in zip(fields, tags) {
    guard field.tag == expectedTag else {
      throw NativeProtocolError.invalidMessage
    }
  }
  return fields
}

private func withInvalidMessage<T>(
  _ operation: () throws -> T
) throws -> T {
  do {
    return try operation()
  } catch {
    throw NativeProtocolError.invalidMessage
  }
}
